core/src/main/java/org/apache/calcite/rex/RexUtil.java

```diff
@@ -2984,6 +2970,9 @@ private static class RexShiftShuttle extends RexShuttle {
     }
   }
 
+  /** Exception to catch when an expression explodes exponentially. */
+  private static class ExponentialExplosionError extends ControlFlowException {}
```

Comment:
It doesn't have to be exponential, sometime even a small polynomial can be too much.
I would call this "PlanTooComplex," since it indicates that optimizing the plan produces a result that is too complex, either at the Rel or at the Rex level (add this to the JavaDoc).